### PR TITLE
Performance Test Set-up (1)

### PR DESCRIPTION
**Pull request by louiseschmidtgen**
## Performance Test

The performance test spins up three lxd containers that join a cluster. Currently, this cluster consists of three control planes. In a follow-up to this PR, kube-burner will run a load-test on the cluster and the resulting metrics are to be displayed as a comment on any future opened PRs.

### Set-up
This PR sets up the harness and gh action for the performance test.

These k8s-dqlite combinations get tested on every test:
| k8s-dqlite | 
|------------| 
| head | 
| base code |
| v1.1.11 | 
| v1.2.0  | 

For simplification, we could opt to run the test on `v1.1.11` and `v1.2.0` only with the published snap (not swapping out the dqlite version).

### Note
Tests are not run as a matrix but in sequence as the collected metrics will be turned into graphs at the end of the workflow.

This test set-up is similar to the k8s-snap integration test. The test is not part of that suite as it is designed as a performance measurement tool of the k8s-dqlite code. It is a tool to directly compare the performance improvements in PRs against the code in master. The test could be extended for microk8s. For those reasons the correct location for this test is in k8s-dqlite.

### CODE REVIEW

File: .github/workflows/performance.yaml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        uses: actions/setup-go@v5
+        with:
+          go-version: "1.21"
+      - name: Install lxd
+        run: |
+          sudo snap refresh lxd --channel 5.21/stable
+          sudo lxd init --auto
+          sudo usermod --append --groups lxd $USER
+          sg lxd -c 'lxc version'
+      - name: Ensure lxd network traffic flows
```
Comment (HomayoonAlimohammadi) [inline .github/workflows/performance.yaml:54]: Is this the right name for this step?

**Review by HomayoonAlimohammadi — APPROVED** [review]
Really amazing job @louiseschmidtgen! I tried to go through it as accurate as I could, and LGTM! It should have been such a huge effort to prepare this PR. Wonderful job!

File: .github/workflows/performance.yaml (hunk 1 above)
Comment (louiseschmidtgen) [reply .github/workflows/performance.yaml:54]: I'll find a better name: docker is dropping some of lxd's network packets which makes the cluster communication impossible.

**Review by eaudetcobello — CHANGES_REQUESTED** [review]
Initial pass

File: .github/workflows/performance.yaml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      - name: Swap out k8s-dqlite and dqlite version for base code ${{ matrix.dqlite }} snap
+        run: |
+          set -o pipefail
+          git fetch origin $BASE_BRANCH
+          git reset --hard $BASE_SHA
+          sed -i 's/^TAG_DQLITE=.*$/TAG_DQLITE=\"${{ matrix.dqlite }}"/' hack/env.sh;
+          make static
+          sudo cp ./bin/static/k8s-dqlite snap-unpack-dir/bin/k8s-dqlite
+          sudo chmod o+r snap-unpack-dir/bin/k8s-dqlite
+          sudo mksquashfs snap-unpack-dir base-code.snap -noappend -comp lzo -no-fragments
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply .github/workflows/performance.yaml]: See Description: Tests are run in sequence as the collected metrics will be turned into graphs at the end of the workflow.

File: test/performance/Readme.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+virtualenv .venv
+. .venv/bin/activate
+pip install 'tox<5'
+```
+
+Further, make sure that you have downloaded the `k8s.snap`:
+
+```bash
+sudo snap download k8s --channel=latest/edge --basename k8s
+```
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply test/performance/Readme.md:24]: This is not the purpose of this workflow. The only thing I need is any snap that does k8s things and swap out its k8s-dqlite with the one that's performance I want to test.

File: test/performance/Readme.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Performance Testing
+
+## Overview
+
+End to end tests are written in Python. They are built on top of a [Harness](./tests/conftest.py) fixture so that they can run on multiple environments like LXD or in the future on the local machine.
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply test/performance/Readme.md:5]: This is a separate test, please read the PR description and the part 2, 3 of the performance test. These are different from the k8s-snap.

File: test/performance/lxd-profile.yaml [hunk 5]
```diff
@@ -0,0 +1,105 @@
+description: "LXD profile for Canonical Kubernetes"
+config:
+  linux.kernel_modules: ip_vs,ip_vs_rr,ip_vs_wrr,ip_vs_sh,ip_tables,ip6_tables,iptable_raw,netlink_diag,nf_nat,overlay,br_netfilter,xt_socket
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply test/performance/lxd-profile.yaml:3]: No, this is an automated workflow and stand-alone test.

File: .github/workflows/performance.yaml [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+          pip install tox
+      - name: Install tox
+        run: pip install tox
+      - name: Install Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: "1.21"
+      - name: Install lxd
+        run: |
+          sudo snap refresh lxd --channel 5.21/stable
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply .github/workflows/performance.yaml]: There used to be an issue with a newer version, let's find out if it is gone.

File: .github/workflows/performance.yaml [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      - name: Install tox
+        run: |
+          pip install --upgrade pip
+          pip install tox
+      - name: Install tox
+        run: pip install tox
+      - name: Install Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: "1.21"
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply .github/workflows/performance.yaml]: Sure we can use 1.22.

File: .github/workflows/performance.yaml [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.8"
+      - name: Install tox
+        run: |
+          pip install --upgrade pip
+          pip install tox
+      - name: Install tox
+        run: pip install tox
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply .github/workflows/performance.yaml]: Thank you :)

File: .github/workflows/performance.yaml [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+      - name: Harden Runner
+        uses: step-security/harden-runner@v2
+        with:
+          egress-policy: audit
+      - name: Checking out repo
+        uses: actions/checkout@v4
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.8"
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply .github/workflows/performance.yaml]: Sure, let's try 3.10 then.

**Review by eaudetcobello — APPROVED** [review]

**Pull request merged**